packages/replay-next/components/ErrorBoundary.tsx

```diff
@@ -1,20 +1,43 @@
 import { captureException } from "@sentry/browser";
-import React, { Component, PropsWithChildren, ReactNode, createElement } from "react";
+import React, { Component, PropsWithChildren, ReactNode } from "react";
 
 import styles from "./ErrorBoundary.module.css";
 
-type ErrorBoundaryState = { error: Error | null };
+type ErrorBoundaryState = { error: Error | null; erroredOnKey: string | number | undefined | null };
 
 type ErrorBoundaryProps = PropsWithChildren<{
   fallback?: ReactNode;
   fallbackClassName?: string;
+  /**
+   * If `resetKey` changes after the ErrorBoundary caught an error, it will reset its state.
+   * Use `resetKey` instead of `key` if you don't want the child components to be recreated
+   * (and hence lose their state) every time the key changes.
+   */
+  resetKey?: string | number;
 }>;
 
 export default class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
-  state: ErrorBoundaryState = { error: null };
+  state: ErrorBoundaryState = { error: null, erroredOnKey: null };
 
   static getDerivedStateFromError(error: Error): ErrorBoundaryState {
-    return { error };
+    return { error, erroredOnKey: null };
+  }
+
+  static getDerivedStateFromProps(
+    props: ErrorBoundaryProps,
+    state: ErrorBoundaryState
+  ): Partial<ErrorBoundaryState> | null {
+    if (state.error) {
+      if (state.erroredOnKey === null) {
+        // we just caught the error and need to save the current resetKey
+        return { erroredOnKey: props.resetKey };
+      }
+      if (props.resetKey !== state.erroredOnKey) {
+        // the resetKey changed after an error was caught so we need to reset the state
+        return { error: null, erroredOnKey: null };
+      }
```

Comment:
I think there's a mismatch between the _optional_ `resetKey` prop (which will be `undefined` by default) and the `erroredOnKey` state (that we initialize to `null`) that will cause unnecessary updates here.

I would make the key optional with a type of `string | null` and then default the value to `null` when the prop wasn't present.

Reply:
> that will cause unnecessary updates here

How?

> I would make the key optional with a type of string | null and then default the value to null when the prop wasn't present.

I deliberately didn't allow the value `null` for `resetKey` because I want `null` to mean that `erroredOnKey` still needs to be set to the current `resetKey` value.

Reply:
> > that will cause unnecessary updates here
>
> How?

```js
// state.erroredOnKey will be "null"
if (state.erroredOnKey === null) {
  // props.resetKey will be "undefined"
  // so this will schedule an unnecessary update with React
  return { erroredOnKey: props.resetKey };
}
```

I think this should be:
```js
if (state.erroredOnKey === null) {
  if (props.resetKey != null) {
    return { erroredOnKey: props.resetKey };
  }
} else if (props.resetKey !== state.erroredOnKey) {
  // the resetKey changed after an error was caught so we need to reset the state
  return { error: null, erroredOnKey: null };
}
```

Reply:
But that would mean the `ErrorBoundary` is reset if `resetKey` changes *except* if it changes to `undefined`. That exception seems unexpected.

Reply:
Also, isn't `getDerivedStateFromProps()` called as part of an update, before rendering, so it wouldn't trigger another rerender?

Reply:
> But that would mean the `ErrorBoundary` is reset if `resetKey` changes *except* if it changes to `undefined`. That exception seems unexpected.

Sorry. I don't understand what you're saying.

I'm suggesting that this whole `getDerivedStateFromProps` should do _nothing_ unless a `resetKey` prop is provided.

> Also, isn't `getDerivedStateFromProps()` called as part of an update, before rendering, so it wouldn't trigger another rerender?

It's still going to require React to restart the render (since it's a _different state value_).

Reply:
I think you're thinking of "render" as literally what's being done in a component's `render` method and I'm referring to "render" as "all of the work React does to render a component– which includes rebasing and diffing state.

> If I have `<ErrorBoundary resetKey={myKey} />`, it will reset if `myKey` changes e.g. from `"foo"` to `"bar"` but not if it changes from `"foo"` to `undefined`.

That's fair. I wasn't considering  the case where we'd pass a `resetKey` and then stop passing one. Hmm... what's the _expected_ behavior there? I could see an argument for why state _shouldn't_ reset in that case.

This is probably well within the "nit" category though so feel free to ignore my input on this.

Reply:
> I'm referring to "render" as "all of the work React does to render a component– which includes rebasing and diffing state.

OK, got it now. Thanks for the entire discussion, I feel like I understand React a little better now.

> what's the expected behavior there? I could see an argument for why state shouldn't reset in that case.

Yeah, it could be argued either way. I'll go with my initial gut feeling that it should also reset in that case.